doc/adr/0016-state-machines-for-UI-state.md

```diff
@@ -0,0 +1,59 @@
+# 0016 State machines for UI state management
+
+Date: Unknown
+
+## Status
+
+Accepted
+
+## Context
+
+Modern (JavaScript-based) User Interfaces have many components and lots of
+concurrent actions going on.  Due to the combination and number of possible
+actions being triggered, the integration between components has become
+increasingly complex over the years -- especially when considering
+asynchronicity and error handling.
+
+The JavaScript/front-end development community has increasingly been using
+[state machines](https://en.wikipedia.org/wiki/Finite-state_machine) and
+[state charts](https://en.wikipedia.org/wiki/UML_state_machine) to solve this
+issue of complexity.
+
+[XState](https://stately.ai/docs/xstate) is the - by far - largest state
+machine engine in the JavaScript ecosystem, featuring over 1 million downloads
+per week (ref: 2023-12-02).  It implements the *state chart* state machine
+type.  XState however is an extensive library with an extensive 'DSL' to
+describe state machines.  There are many alternatives implementing state charts
+or state machines that are less extensive.  One such example is [robot3](https://thisrobot.life/)
+which is extremely small, supports sub-state machines and features integration
+for React, PReact and Lit.  Its DSL is similarly small and thereby easy to
+read.  None of the integrations are of interest for LedgerSMB, but they
+indicate interest from a wider community.
+
+Considering that with its move to Vue, LedgerSMB is going in the same
+direction as other projects -- featuring a heavy (JavaScript-based) client
```

Comment:
s/heavy/rich/ ? fat?

I think we should strive to keep the front end light, not heavy -- but add rich functionality to improve the user experience. https://infrequently.org/2023/02/the-market-for-lemons/ (author is the original author of Dojo, Alex Russell -- see footnote [1] where he does mention Vue in the list of frameworks that care about performance/small footprint).

So we should emphasize trying to stick as close to vanilla js and css as we can, and use tools like Vue and Pinia as added sugar...

Reply:
Added.